eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovSpec.hs

```diff
@@ -55,6 +60,809 @@ spec =
+                    ]
+                , E []
+                ]
+
+          forest <- getProposals
+          passNEpochs 5
+          forest' <- getProposals
+          forest' `shouldBe` forest
+          passEpoch
+          forest'' <- getProposals
+          forest'' `shouldBe` def
+        it "Subtrees are pruned when proposals expire" $ do
+          modifyPParams $ ppGovActionLifetimeL .~ EpochInterval 4
+          p1 <- submitInitConstitutionGovAction
+          passNEpochs 3
+          N
+            p11
+            [ N p111 []
+              , N p112 []
+              ] <-
+            submitConstitutionGovActionTree
+              (SJust p1)
+              $ E
+                [ E []
+                , E []
+                ]
+          N p2 [N p21 []] <-
+            submitConstitutionGovActionTree
+              SNothing
+              $ E
+                [ E []
+                ]
+          [ I []
+            , I []
+            , I []
+            , I
+                [ J
+                    p1'
+                    [ J
+                        p11'
+                        [ J p111' []
+                          , J p112' []
+                          ]
+                      ]
+                  , J p2' [J p21' []]
+                  ]
+            ] <-
+            getProposalsForest
+          SJust p1 `shouldBe` p1'
+          SJust p2 `shouldBe` p2'
+          SJust p11 `shouldBe` p11'
+          SJust p21 `shouldBe` p21'
+          SJust p111 `shouldBe` p111'
+          SJust p112 `shouldBe` p112'
+          passNEpochs 3
+          [ I []
+            , I []
+            , I []
+            , I
+                [ J p2'' [J p21'' []]
+                  ]
+            ] <-
+            getProposalsForest
+          SJust p2 `shouldBe` p2''
+          SJust p21 `shouldBe` p21''
+        it "Subtrees are pruned when competing proposals are enacted" $ do
+          (dRep, committeeMember) <- electBasicCommittee
+          [ N
+              p1
+              [ N
+                  p11
+                  [ N p111 []
+                    , N p112 []
+                    ]
+                ]
+            , N p2 [N p21 []]
+            ] <-
+            submitConstitutionGovActionForest
+              SNothing
+              [ E
+                  [ E
+                      [ E []
+                      , E []
+                      ]
+                  ]
+              , E
+                  [ E []
+                  ]
+              ]
+          [ I []
+            , I []
+            , J _ [] -- Due to the committee enactment
+            , I
+                [ J
+                    p1'
+                    [ J
+                        p11'
+                        [ J p111' []
+                          , J p112' []
+                          ]
+                      ]
+                  , J p2' [J p21' []]
+                  ]
+            ] <-
+            getProposalsForest
+          SJust p1 `shouldBe` p1'
```

Comment:
I think you could also compare the whole tree at once instead of comparing each node manually

Reply:
I'm not sure how to do this, since the trees are being constructed in parts and we would need to reconstruct the tree any way to be able to compare the whole with another whole. Can you give me an example of how you mean we could do this?

Reply:
I see now that the problem is that one of the trees is polymorphic over `StrictMaybe` of `GovActionId` and the other is just `GovActionId`. You'd probably have to map over the second tree with `SJust` and then you could do a direct comparison on the whole tree.

Reply:
> I think you could also compare the whole subtree at once instead of comparing each node manually, should save some lines.

> .. You'd probably have to map over the second tree ...

@aniketd That is exactly what I was suggesting you during our 1on1 as well

Reply:
Done as much as I could do :+1: